File: 练习/MyWeiBo/MyWeiBo/Class/Base/View/HJTabBar.m
//
//  HJTabBar.m
//  MyWeiBo
//
//  Created by M-coppco on 16/5/1.
//  Copyright © 2016年 XHJ. All rights reserved.
//

#import "HJTabBar.h"

@interface HJTabBar ()
@property (nonatomic, strong)UIButton  *addButton;
@end

@implementation HJTabBar

- (instancetype)init {
    self  = [super init];
    if (self) {
        self.addButton = [UIButton buttonWithType:(UIButtonTypeCustom)];
        [self.addButton setBackgroundImage:[UIImage imageNamed:@"tabbar_compose_button"] forState:(UIControlStateNormal)];
        [self.addButton setBackgroundImage:[UIImage imageNamed:@"tabbar_compose_button_highlighted"] forState:(UIControlStateHighlighted)];
        [self.addButton setImage:[UIImage imageNamed:@"tabbar_compose_icon_add"] forState:(UIControlStateNormal)];
        [self.addButton setImage:[UIImage imageNamed:@"tabbar_compose_icon_add_highlighted"] forState:(UIControlStateNormal)];
        self.addButton.size = self.addButton.currentBackgroundImage.size;
        [self.addButton addTarget:self action:@selector(buttonClick) forControlEvents:(UIControlEventTouchUpInside)];
        [self addSubview:self.addButton];
        
    }
    return self;
}
- (void)buttonClick {
    if (self.click) {
        self.click();
    }
}
- (void)layoutSubviews {
    [super layoutSubviews];
    
    //设置加号按钮的位置
    self.addButton.centerX = self.width * 0.5;
    self.addButton.centerY = self.height * 0.5;
    
    CGFloat tabbarButtonW = self.width / 5;
    CGFloat tabbarButtonIndex = 0;
    
    for (int i = 0; i < self.subviews.count; i++) {
        UIView *view = self.subviews[i];
        Class class = NSClassFromString(@"UITabBarButton");
        
        if ([view isKindOfClass:class]) {
            view.width = tabbarButtonW;
            view.x = tabbarButtonW * tabbarButtonIndex;
            tabbarButtonIndex++;
            
            if (tabbarButtonIndex == 2) {
                tabbarButtonIndex++;
            }
            
        }
    }
}

@end

File: 小码哥大神5期练习/6、Swift版本-百思不得姐/Best Not Elder Sister/Best Not Elder Sister/Class/HJTool/HJExtension/UIView/HJTabBar.swift
//
//  HJTabBar.swift
//  Best Not Elder Sister
//
//  Created by M-coppco on 16/7/24.
//  Copyright © 2016年 M-coppco. All rights reserved.
//

import UIKit

class HJTabBar: UITabBar {
    
    private var plusClick: (HJTabBar) -> Void?
    
    private lazy var publishButton: UIButton = {
            let button = UIButton(type: .Custom)
            button.setImage(UIImage(named: "tabBar_publish_icon"), forState: .Normal)
            button.setImage(UIImage(named: "tabBar_publish_click_icon"), forState: .Highlighted)
            button.addTarget(self, action: "publish", forControlEvents: .TouchUpInside)
        return button
    }()
    
    init(closure: (tabBar: HJTabBar) -> Void) {
        self.plusClick = closure
        super.init(frame: CGRectZero)
        self.addSubview(publishButton)
        self.backgroundImage = UIImage(named: "tabbar-light")
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        var index: Int = 0
        for view in self.subviews {
            if  view.dynamicType == NSClassFromString("UITabBarButton") {
                if index < 2{
                    view.frame = CGRectMake(CGFloat(index) * self.hj_width / 5, 0, self.hj_width / 5, self.hj_height)
                } else {
                    view.frame = CGRectMake(CGFloat(index + 1) * self.hj_width / 5, 0, self.hj_width / 5, self.hj_height)
                }
                index++
            }
        }
        
        publishButton.frame = CGRectMake(2 * self.hj_width / 5, 0, self.hj_width / 5, self.hj_height)
    }
    func publish() {
        plusClick(self)
    }
}
